pkg/ee/validation/machine/resourcequota.go

```diff
@@ -0,0 +1,155 @@
+		if err != nil {
+			return fmt.Errorf("error getting fake quota req: %w", err)
+		}
+	default:
+		// TODO skip for now, when all providers are added, throw error
+		log.Debugf("provider %q not supported", config.CloudProvider)
+		return nil
+	}
+
+	// TODO Get quota and usage from ResourceQuota when its implemented
+	quota, currentUsage, err := getResourceQuota()
+	if err != nil {
+		return fmt.Errorf("failed to get resource quota: %w", err)
+	}
+
+	// add requested resources to current usage and compare
+	combinedUsage := NewResourceQuota(currentUsage.cpu, currentUsage.mem, currentUsage.storage)
+	combinedUsage.Cpu().Add(*quotaReq.Cpu())
+	combinedUsage.Memory().Add(*quotaReq.Memory())
+	combinedUsage.Storage().Add(*quotaReq.Storage())
+
+	if quota.Cpu().Cmp(*combinedUsage.Cpu()) < 0 {
+		log.Debugf("requested CPU %q would exceed current quota (quota/used %q/%q)",
+			quotaReq.Cpu(), quota.Cpu(), currentUsage.Cpu())
+		return fmt.Errorf("requested CPU %q would exceed current quota (quota/used %q/%q)",
+			quotaReq.Cpu(), quota.Cpu(), currentUsage.Cpu())
+	}
+
+	if quota.Memory().Cmp(*combinedUsage.Memory()) < 0 {
+		log.Debugf("requested Memory %q would exceed current quota (quota/used %q/%q)",
+			quotaReq.Memory(), quota.Memory(), currentUsage.Memory())
+		return fmt.Errorf("requested Memory %q would exceed current quota (quota/used %q/%q)",
+			quotaReq.Memory(), quota.Memory(), currentUsage.Memory())
+	}
+
+	if quota.Storage().Cmp(*combinedUsage.Storage()) < 0 {
+		log.Debugf("requested Storage %q would exceed current quota (quota/used %q/%q)",
+			quotaReq.Storage(), quota.Storage(), currentUsage.Storage())
+		return fmt.Errorf("requested Storage %q would exceed current quota (quota/used %q/%q)",
+			quotaReq.Storage(), quota.Storage(), currentUsage.Storage())
+	}
+
+	return nil
+}
+
+// TODO we should get it from the resourceQuota CRD for the project, for now just some hardcoded values for tests.
+func getResourceQuota() (*ResourceQuota, *ResourceQuota, error) {
+	cpu, err := resource.ParseQuantity("5")
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing quantity: %w", err)
+	}
+	cpuUsed, err := resource.ParseQuantity("3")
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing quantity: %w", err)
+	}
+
+	mem, err := resource.ParseQuantity("5G")
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing quantity: %w", err)
+	}
+	memUsed, err := resource.ParseQuantity("3G")
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing quantity: %w", err)
+	}
+
+	storage, err := resource.ParseQuantity("100G")
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing quantity: %w", err)
+	}
+	storageUsed, err := resource.ParseQuantity("60G")
+	if err != nil {
+		return nil, nil, fmt.Errorf("error parsing quantity: %w", err)
+	}
+
+	return NewResourceQuota(cpu, mem, storage),
+		NewResourceQuota(cpuUsed, memUsed, storageUsed), nil
+}
+
+type ResourceQuota struct {
```

Comment:
This struct seems to be used not just for quotas, but also to express the current consumption. So the name for this struct might be a bit unfortunate. What about `ResourceDetails` maybe?

Reply:
What about using `corev1.ResourceList`? That one looks super similar, maybe similar enough?

Reply:
My first thought and implementation was with `corev1.ResourceList`, but it has some additional things like `ResourceEphemeralStorage` and Pods() which I thought could make it confusing. So I decided on something simpler